node/src/components/storage.rs

```diff
@@ -895,18 +904,57 @@ impl Storage {
         }))
     }
 
-    // Retrieves a block header to handle a network request.
-    pub fn read_block_header_and_finality_signatures_by_height(
+    /// Retrieves a block by hash.
+    pub fn read_block(&self, block_hash: &BlockHash) -> Result<Option<Block>, LmdbExtError> {
+        self.get_single_block(&mut self.env.begin_ro_txn()?, block_hash)
+    }
+
+    /// Retrieves a block header by height.
+    /// Returns `None` if they are less than the fault tolerance threshold, or if the block is from
+    /// before the most recent emergency upgrade.
+    pub fn read_block_header_and_sufficient_finality_signatures_by_height(
         &self,
         height: u64,
+        genesis_validator_weights: &BTreeMap<PublicKey, U512>,
+        finality_threshold_fraction: Ratio<u64>,
+        last_emergency_restart: Option<EraId>,
     ) -> Result<Option<BlockHeaderWithMetadata>, Error> {
         let mut txn = self.env.begin_ro_txn()?;
-        let maybe_block_header_and_finality_signatures =
-            self.get_block_header_and_metadata_by_height(&mut txn, height)?;
+        let maybe_block_header_and_finality_signatures = self
+            .get_block_header_and_sufficient_finality_signatures_by_height(
+                &mut txn,
+                height,
+                genesis_validator_weights,
+                finality_threshold_fraction,
+                last_emergency_restart,
+            )?;
         drop(txn);
         Ok(maybe_block_header_and_finality_signatures)
     }
 
+    /// Retrieves a block by height.
+    /// Returns `None` if they are less than the fault tolerance threshold, or if the block is from
+    /// before the most recent emergency upgrade.
+    pub fn read_block_and_sufficient_finality_signatures_by_height(
+        &self,
+        height: u64,
+        genesis_validator_weights: &BTreeMap<PublicKey, U512>,
+        finality_threshold_fraction: Ratio<u64>,
+        last_emergency_restart: Option<EraId>,
+    ) -> Result<Option<BlockWithMetadata>, Error> {
+        let mut txn = self.env.begin_ro_txn()?;
+        let maybe_block_and_finality_signatures = self
+            .get_block_and_sufficient_finality_signatures_by_height(
+                &mut txn,
+                height,
+                genesis_validator_weights,
+                finality_threshold_fraction,
+                last_emergency_restart,
+            )?;
+        drop(txn);
+        Ok(maybe_block_and_finality_signatures)
+    }
```

Comment:
These belong to the `LinearChain` component in my opinion, and not storage. Storage shouldn't have _any_ domain logic, it should only handle retrieving and storing data.

Reply:
While it is true that the storage component now contains domain logic, involving the linear chain has significant downsides.

1. It does not access LMDB in a transactional manner. We are losing the A in ACID. The code in this PR is transactional.

2. The JSON-RPC accesses the Storage component directly rather than using the `LinearChain`. It does not require sufficient finality signatures as a requirement. While it _could_, that would require creating several different message types and extending various handlers.

3. There are 6 different fetchers. While two of them routed to the linear chain, the rest do not.  All of the fetcher requests are now handled uniformly [here][1].

[1]: https://github.com/casper-network/casper-node/pull/1728/files#diff-212250b2151062263c01ec4511602e2ada0b8a83dd02b8727589737f3d3312ebR366-R427